common/src/main/java/com/wynntils/core/net/Download.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.net;
+
+import com.wynntils.core.WynntilsMod;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.http.HttpHeaders;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import org.apache.commons.io.FileUtils;
+
+public class Download extends NetResult {
+    private static final Map<HttpRequest, CompletableFuture<HttpResponse<Path>>> DOWNLOAD_FUTURES = new HashMap<>();
+
+    private final File localFile;
+
+    // Saved since we might need to get timestamps from the HttpResponse
+    private CompletableFuture<HttpResponse<Path>> httpResponse = null;
+
+    public Download(File localFile) {
+        super(null); // Only use cached file
+        this.localFile = localFile;
+    }
+
+    public Download(File localFile, HttpRequest request) {
+        super(request);
+        this.localFile = localFile;
+    }
+
+    public long getResponseTimestamp() {
+        if (httpResponse == null) {
+            // We have either not yet made the request, or we have read from the cache
+            // In either case, the best we can do is to return the current time
+            return System.currentTimeMillis();
+        }
+
+        try {
+            HttpHeaders headers = httpResponse.get().headers();
+            OptionalLong a = headers.firstValueAsLong("timestamp");
+            if (a.isEmpty()) return System.currentTimeMillis();
+            return a.getAsLong();
+        } catch (InterruptedException | ExecutionException e) {
+            return System.currentTimeMillis();
+        }
+    }
+
+    protected CompletableFuture<InputStream> getInputStreamFuture() {
+        if (request == null) {
+            // File is already in downloaded, just read from the cache
```

Comment:
What is we have a request and a local file? It is likely handled fine but I have not yet understand the code as a whole

Reply:
We will always have a local file, as in a `File` object. (That does not mean that there exists a corresponding file with data content on the disk). This is where the download is/will be cached. The primary difference between Download and ApiResult is that a Download always downloads to a file first, and then reads from the file to handle the content.

If `request` is null, then the NetManager has established that there already exists a file at the location pointed to by the local file `File` object, and that it has a matching md5 sum, so we do not need to download anything. Instead, we can just read the contents of that file directly, and do the async handling/parsing of that file.

Reply:
Ah, now I think I understand what you mean. It is the responsibility of the NetManager to determine if the cached file is usable, which it will be deemed to be if we know the md5 sum for that URL, and the md5 sum of the local file matches.

There were some efforts in the old implementation to use the cache "as backup", even without looking at md5, in case of network failures. I don't think it really worked as intended. But the idea is good. I'd like to implement something similar here, but it is non-trivial. You'd have to hook in at the right place to figure out there is a network failure but a local file that might do. And you should probably inform the caller that it is operating on a stale cache.